packages/backend-common/src/reading/BitbucketUrlReader.ts

```diff
@@ -138,8 +150,42 @@ export class BitbucketUrlReader implements UrlReader {
     throw new Error(message);
   }
 
-  readTree(): Promise<ReadTreeResponse> {
-    throw new Error('BitbucketUrlReader does not implement readTree');
+  async readTree(
+    url: string,
+    options?: ReadTreeOptions,
+  ): Promise<ReadTreeResponse> {
+    const gitUrl: parseGitUri.GitUrl = parseGitUri(url);
+    const { name: repoName, owner, ref, protocol, resource, filepath } = gitUrl;
+
+    const isHosted = resource === 'bitbucket.org';
+
+    if (isHosted && !ref) {
+      // TODO(freben): We should add support for defaulting to the default branch
+      throw new InputError(
+        'Bitbucket URL must contain branch to be able to fetch tree',
+      );
+    }
+
+    const archiveUrl = isHosted
+      ? `${protocol}://${resource}/${owner}/${repoName}/get/${ref}.zip`
+      : `${protocol}://${resource}/projects/${owner}/repos/${repoName}/archive?format=zip`;
```

Comment:
Oh and also, `path=`, so that we download only the part of the repo necessary instead of the entire repo.

This would make the two parts sufficiently different to warrant being broken apart into two functions.

Reply:
I just want to say - the above are not mandatory. If your current solution works and it gets approved, then it's perfectly fine to have it merged and then you could make a small issue pointing to these lines asking for future improvement (since it's basically just an optimization). Then we can add good-first-issue and help-wanted labels on it and the community might help out.

Reply:
I can look into these suggestions!

Reply:
`getBitbucketDownloadUrl` is implemented in `packages/integration/src/bitbucket/core.ts` with the suggested parameters for the case where the repository is on Bitbucket Server.